src/Workspaces/Core/Portable/ObsoleteSymbol/AbstractObsoleteSymbolService.cs

```diff
@@ -0,0 +1,176 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageService;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.ObsoleteSymbol;
+
+internal abstract class AbstractObsoleteSymbolService(int? dimKeywordKind) : IObsoleteSymbolService
+{
+    private readonly int? _dimKeywordKind = dimKeywordKind;
+
+    protected virtual void ProcessDimKeyword(ArrayBuilder<TextSpan> result, SemanticModel semanticModel, SyntaxToken token, CancellationToken cancellationToken)
+    {
+        // Take no action by default
+    }
+
+    public async Task<ImmutableArray<TextSpan>> GetLocationsAsync(Document document, ImmutableArray<TextSpan> textSpans, CancellationToken cancellationToken)
+    {
+        var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+        var compilation = await document.Project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+        using var _2 = ArrayBuilder<TextSpan>.GetInstance(out var result);
```

Comment:
nit: obsolete is likely so rare that usign TempArray is likely ok :)

Reply:
➡️ Kept `ArrayBuilder<T>` so I didn't need to reimplement `RemoveDuplicates`, but rewrote the code to avoid taking the builder from the pool in the overwhelming majority case where no obsolete symbols are used in the classification region.